massa-hash/src/hash_xof.rs

```diff
@@ -0,0 +1,157 @@
+use std::ops::{BitXor, BitXorAssign};
+
+use massa_serialization::{Deserializer, SerializeError, Serializer};
+use nom::{
+    error::{context, ContextError, ParseError},
+    IResult,
+};
+
+/// Extended Hash
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct HashXof<const SIZE: usize>(pub [u8; SIZE]);
+
+impl<const SIZE: usize> HashXof<SIZE> {
+    /// From bytes
+    pub fn from_bytes(bytes: &[u8; SIZE]) -> Self {
+        HashXof(*bytes)
+    }
+
+    /// Transform into bytes
+    pub fn to_bytes(&self) -> &[u8; SIZE] {
+        &self.0
+    }
+
+    /// Compute from raw data
+    pub fn compute_from(data: &[u8]) -> HashXof<SIZE> {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(data);
+        let mut hash = [0u8; SIZE];
+        let mut output_reader = hasher.finalize_xof();
+        output_reader.fill(&mut hash);
+        HashXof(hash)
+    }
+
+    /// Compute from key and value
+    pub fn compute_from_kv(key: &[u8], value: &[u8]) -> HashXof<SIZE> {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&(key.len() as u64).to_be_bytes());
+        hasher.update(key);
+        hasher.update(value);
+        let mut hash = [0u8; SIZE];
+        let mut output_reader = hasher.finalize_xof();
+        output_reader.fill(&mut hash);
+        HashXof(hash)
+    }
+
+    /// Serialize a Hash using `bs58` encoding with checksum.
+    /// Motivations for using base58 encoding:
+    ///
+    /// base58_check is like base64 but-
+    /// * fully standardized (no = vs /)
+    /// * no weird characters (eg. +) only alphanumeric
+    /// * ambiguous letters combined (eg. O vs 0, or l vs 1)
+    /// * contains a checksum at the end to detect typing errors
+    ///    
+    pub fn to_bs58_check(&self) -> String {
+        bs58::encode(self.0).with_check().into_string()
+    }
+}
+
+impl<const SIZE: usize> BitXorAssign for HashXof<SIZE> {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        *self = *self ^ rhs;
+    }
+}
+
+impl<const SIZE: usize> BitXor for HashXof<SIZE> {
```

Comment:
Maybe explain / refer to the discussion here and on the BitXorAssign the security considerations for the SIZE const?

Reply:
I don't know which discussion you speak about for the xor operator

Reply:
https://github.com/massalabs/massa/discussions/3852#discussioncomment-6188158

Reply:
Added.